// Copyright 2024 Notedown Authors
//
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
//     http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.

syntax = "proto3";

package extract;

service Extract {
    rpc Extract(ExtractRequest) returns (ExtractResponse) {}
}

enum EmbeddingsModel {
    NVIDIA_EMBED_V2 = 0;
}

enum SpeechToTextModel {
    OPENAI_WHISPER_LARGE_V3 = 0;
}

message ExtractArticleRequest {
    string url = 1;
}

message ExtractVideoRequest {
    string url = 1;
    SpeechToTextModel speech_to_text_model = 2;
}

message ExtractRequest {
    oneof type {
        ExtractArticleRequest article = 1;
        ExtractVideoRequest video = 2;
    }
}

enum Format {
    ARTICLE = 0;
    VIDEO = 1;
    PODCAST = 2;
    BOOK = 3;
    COURSE = 4;
}

message ExtractResponse {
    string title = 1;
    Format format = 2;
    bytes content = 3;
}
